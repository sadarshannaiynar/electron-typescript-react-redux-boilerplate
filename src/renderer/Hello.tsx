import * as React from 'react';

interface Props {
  name: string;
}

interface State {
  name: string;
}

export class HelloComponent extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      name: props.name,
    };
    this.handleNameChange = this.handleNameChange.bind(this);
  }

  componentWillReceiveProps(nextProps: Props) {
    this.setState({
      name: nextProps.name,
    });
  }

  handleNameChange(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({
      name: event.target.value,
    });
  }

  render() {
    return (
      <div>
        <h4 className="name">{this.state.name}</h4>
        <input type="text" onChange={this.handleNameChange} />
      </div>
    );
  }
}
