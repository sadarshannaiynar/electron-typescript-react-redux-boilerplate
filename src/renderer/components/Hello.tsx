import * as React from 'react';
import { IHelloContainerDispatch, IHelloContainerState } from '../containers/HelloContainer';

export type HelloComponentProps = IHelloContainerState & IHelloContainerDispatch;

interface IHelloComponentState {
  value: number;
}

export class HelloComponent extends React.Component<HelloComponentProps, IHelloComponentState> {
  constructor(props: HelloComponentProps) {
    super(props);
    this.state = {
      value: props.value,
    };
    this.handleValueChange = this.handleValueChange.bind(this);
  }

  componentWillReceiveProps(nextProps: HelloComponentProps) {
    this.setState({
      value: nextProps.value,
    });
  }

  handleValueChange(event: React.ChangeEvent<HTMLInputElement>) {
    const numericValue = parseInt(event.target.value, 10);
    this.setState({
      value: numericValue,
    });
    this.props.setValue(numericValue);
  }

  render() {
    return (
      <div>
        <div>Hello World!</div>
        <h4 className="value">{this.state.value}</h4>
        <button className="increment-btn" onClick={this.props.increment}>Increment</button>
        &nbsp;
        <button className="decrement-btn" onClick={this.props.decrement}>Decrement</button>
        <br /><br />
        Type a value:
        <input type="text" onChange={this.handleValueChange} />
        <br /><br />
        <button className="execute-ls-btn" onClick={this.props.executeLs}>Execute LS Command</button>
        <br /><br />
        <div className="output">Output: {this.props.output}</div>
        <br /><br />
        <div className="error">Error: {this.props.error}</div>
      </div>
    );
  }
}
