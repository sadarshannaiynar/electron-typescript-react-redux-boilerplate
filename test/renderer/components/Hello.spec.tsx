import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import { HelloComponent, HelloComponentProps } from '../../../src/renderer/components/Hello';

interface IWrapperAndMocks {
  wrapper: ShallowWrapper;
  mockSetValue: jest.Mock;
  mockIncrement: jest.Mock;
  mockDecrement: jest.Mock;
  mockExecuteLs: jest.Mock;
}

const getWrapperAndMocks = (value: number, output: string, error: string): IWrapperAndMocks => {
  const mockSetValue = jest.fn();
  const mockIncrement = jest.fn();
  const mockDecrement = jest.fn();
  const mockExecuteLs = jest.fn();

  const props: HelloComponentProps = {
    error,
    output,
    value,
    decrement: mockDecrement,
    executeLs: mockExecuteLs,
    increment: mockIncrement,
    setValue: mockSetValue,
  };

  const wrapper = shallow(<HelloComponent {...props} />);

  return {
    mockDecrement,
    mockExecuteLs,
    mockIncrement,
    mockSetValue,
    wrapper,
  };
};

describe('<HelloComponent /> Tests', () => {
  describe('Rendering Tests', () => {
    const { wrapper } = getWrapperAndMocks(0, 'Output', 'Error');

    test('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('should render the Hello World! div', () => {
      expect(wrapper.find('div > div').first().text()).toStrictEqual('Hello World!');
    });

    test('should render the value', () => {
      expect(wrapper.find('.value').exists()).toBeTruthy();
      expect(wrapper.find('.value').text()).toStrictEqual('0');
    });

    test('should render the output', () => {
      expect(wrapper.find('.output').exists()).toBeTruthy();
      expect(wrapper.find('.output').text()).toStrictEqual('Output: Output');
    });

    test('should render the error', () => {
      expect(wrapper.find('.error').exists()).toBeTruthy();
      expect(wrapper.find('.error').text()).toStrictEqual('Error: Error');
    });

    test('should render the increment-btn', () => {
      expect(wrapper.find('.increment-btn').first().text()).toStrictEqual('Increment');
    });

    test('should render the decrement-btn', () => {
      expect(wrapper.find('.decrement-btn').first().text()).toStrictEqual('Decrement');
    });

    test('should render the execute-ls-btn', () => {
      expect(wrapper.find('.execute-ls-btn').first().text()).toStrictEqual('Execute LS Command');
    });

    test('should render the input field', () => {
      expect(wrapper.find('input').exists()).toBeTruthy();
    });
  });

  describe('Lifecycle Tests', () => {
    const { wrapper } = getWrapperAndMocks(1, '', '');

    test('should call componentWillReceiveProps when new props are received', () => {
      expect(wrapper.state('value')).toStrictEqual(1);
      wrapper.setProps({ value: 5 });
      expect(wrapper.state('value')).toStrictEqual(5);
    });
  });

  describe('Functionality Tests', () => {
    test('should call onChange of input when value changes', () => {
      const { wrapper, mockSetValue } = getWrapperAndMocks(1, '', '');

      expect(wrapper.state('value')).toStrictEqual(1);
      wrapper.find('input').first().simulate('change', { target: { value: '4' } });
      expect(wrapper.state('value')).toStrictEqual(4);
      expect(mockSetValue).toHaveBeenCalledTimes(1);
      expect(mockSetValue).toHaveBeenLastCalledWith(4);
    });

    test('should call increment function when increment-btn is clicked', () => {
      const { wrapper, mockIncrement } = getWrapperAndMocks(1, '', '');

      wrapper.find('.increment-btn').first().simulate('click');
      expect(mockIncrement).toHaveBeenCalledTimes(1);
      expect(mockIncrement).toHaveBeenLastCalledWith();
    });

    test('should call decrement function when decrement-btn is clicked', () => {
      const { wrapper, mockDecrement } = getWrapperAndMocks(1, '', '');

      wrapper.find('.decrement-btn').first().simulate('click');
      expect(mockDecrement).toHaveBeenCalledTimes(1);
      expect(mockDecrement).toHaveBeenLastCalledWith();
    });

    test('should call executeLs function when execute-ls-btn is clicked', () => {
      const { wrapper, mockExecuteLs } = getWrapperAndMocks(3, '', '');

      wrapper.find('.execute-ls-btn').first().simulate('click');
      expect(mockExecuteLs).toHaveBeenCalledTimes(1);
      expect(mockExecuteLs).toHaveBeenLastCalledWith();
    });
  });
});
