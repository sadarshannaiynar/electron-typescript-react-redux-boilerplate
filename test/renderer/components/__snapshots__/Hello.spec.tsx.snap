// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HelloComponent /> Tests Rendering Tests should match the snapshot 1`] = `
<div>
  <div>
    Hello World!
  </div>
  <h4
    className="value"
  >
    0
  </h4>
  <button
    className="increment-btn"
    onClick={[MockFunction]}
  >
    Increment
  </button>
  Â 
  <button
    className="decrement-btn"
    onClick={[MockFunction]}
  >
    Decrement
  </button>
  <br />
  <br />
  Type a value:
  <input
    onChange={[Function]}
    type="text"
  />
  <br />
  <br />
  <button
    className="execute-ls-btn"
    onClick={[MockFunction]}
  >
    Execute LS Command
  </button>
  <br />
  <br />
  <div
    className="output"
  >
    Output: 
    Output
  </div>
  <br />
  <br />
  <div
    className="error"
  >
    Error: 
    Error
  </div>
</div>
`;
